### PR TITLE
Removed `anyhow`

### LINKED ISSUE #8140: Reduce/remove the use of  `anyhow` in the engine
I'd really like for us to reduce our use of `anyhow`. IMO it's a crate that end-users should be using, not the libraries they're depending on.

_Originally posted by @james7132 in https://github.com/bevyengine/bevy/pull/8065#discussion_r1142590436_

### PR DESCRIPTION
**Pull request by bushrat011899**
# Objective

- Fixes #8140

## Solution

- Added Explicit Error Typing for `AssetLoader` and `AssetSaver`, which were the last instances of `anyhow` in use across Bevy.

---

## Changelog

- Added an associated type `Error` to `AssetLoader` and `AssetSaver` for use with the `load` and `save` methods respectively.
- Changed `ErasedAssetLoader` and `ErasedAssetSaver` `load` and `save` methods to use `Box<dyn Error + Send + Sync + 'static>` to allow for arbitrary `Error` types from the non-erased trait variants. Note the strict requirements match the pre-existing requirements around `anyhow::Error`.

## Migration Guide

- `anyhow` is no longer exported by `bevy_asset`; Add it to your own project (if required).
- `AssetLoader` and `AssetSaver` have an associated type `Error`; Define an appropriate error type (e.g., using `thiserror`), or use a pre-made error type (e.g., `anyhow::Error`). Note that using `anyhow::Error` is a drop-in replacement.
- `AssetLoaderError` has been removed; Define a new error type, or use an alternative (e.g., `anyhow::Error`)
- All the first-party `AssetLoader`'s and `AssetSaver`'s now return relevant (and narrow) error types instead of a single ambiguous type; Match over the specific error type, or encapsulate (`Box<dyn>`, `thiserror`, `anyhow`, etc.)

## Notes

A simpler PR to resolve this issue would simply define a Bevy `Error` type defined as `Box<dyn std::error::Error + Send + Sync + 'static>`, but I think this type of error handling should be discouraged when possible. Since only 2 traits required the use of `anyhow`, it isn't a substantive body of work to solidify these error types, and remove `anyhow` entirely. End users are still encouraged to use `anyhow` if that is their preferred error handling style. Arguably, adding the `Error` associated type gives more freedom to end-users to decide whether they want more or less explicit error handling (`anyhow` vs `thiserror`).

As an aside, I didn't perform any testing on Android or WASM. CI passed locally, but there may be mistakes for those platforms I missed.

### CODE REVIEW

File: crates/bevy_render/src/render_resource/shader.rs [hunk 1]
```diff
@@ -232,15 +232,24 @@ impl From<&Source> for naga_oil::compose::ShaderType {
 #[derive(Default)]
 pub struct ShaderLoader;
 
+#[derive(Debug, Error)]
+pub enum ShaderLoaderError {
+    #[error("Could load shader: {0}")]
```
Comment (JMS55) [inline crates/bevy_render/src/render_resource/shader.rs:237]:
```suggestion
    #[error("Could not load shader: {0}")]
```

**Review by killercup — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/render_resource/shader.rs (hunk 1 above)
Comment (bushrat011899) [reply crates/bevy_render/src/render_resource/shader.rs]: Whoops! Thanks for the pickup!

File: crates/bevy_asset/src/lib.rs [hunk 2]
```diff
@@ -462,24 +463,40 @@ mod tests {
     #[derive(Default)]
     struct CoolTextLoader;
 
+    #[derive(Error, Debug)]
+    enum CoolTextLoaderError {
+        #[error("Could not load dependency: {dependency}")]
+        CannotLoadDependency { dependency: AssetPath<'static> },
+        #[error("A RON error occurred during loading")]
+        RONSpannedError(#[from] ron::error::SpannedError),
```
[the remark this answers is not in the document]
Comment (bushrat011899) [reply crates/bevy_asset/src/lib.rs]: Yeah that does look nicer, thanks!

File: crates/bevy_asset/src/lib.rs [hunk 3]
```diff
@@ -462,24 +463,40 @@ mod tests {
...
+        #[error("An IO error occurred during loading")]
+        IO(#[from] std::io::Error),
```
[the remark this answers is not in the document]
Comment (bushrat011899) [reply crates/bevy_asset/src/lib.rs]: As before, thanks!

File: crates/bevy_asset/src/lib.rs [hunk 4]
```diff
@@ -462,24 +463,40 @@ mod tests {
     #[derive(Default)]
     struct CoolTextLoader;
 
+    #[derive(Error, Debug)]
```
[the remark this answers is not in the document]
Comment (bushrat011899) [reply crates/bevy_asset/src/lib.rs:466]: For the new error types I'll add `#[non_exhaustive]`, but I'll leave it off for any that already existed, unless others agree they should be changed to include that attribute.

**Review by killercup — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]
This is much nicer. Thank you!

**Comment by bushrat011899** [comment]
Looks like I missed one docstring in the `bevy_gltf` package (which previously didn't require documentation). I've merged `main` back into this PR and resolved that issue.

**Comment by cart** [comment]
These AssetLoadError::CannotLoadDependency errors don't make sense to me:

1. We're dropping the actual loader error contents, making the error message effectively useless.
2. Not all asset loads are dependencies, but we're returning that error for every asset load failure.

**Comment by cart** [comment]
We're also no longer including meta deserialization failures / throwing them in the generic CannotLoadDependency bucket.

**Comment by cart** [comment]
I'm just going to fix this in Multiple Asset Sources as it touches this code.

**Comment by Roryl-c** [comment]
> AssetLoader and AssetSaver have an associated type Error; Define an appropriate error type (e.g., using thiserror), or use a pre-made error type (e.g., anyhow::Error). Note that using anyhow::Error is a drop-in replacement.  

Not sure if a comment is the right place to put this, but this section of this PR's migration guide is incorrect. `anyhow::Error` does not implement `Error`, which is required by these associated types - so it does not work as a drop-in replacement.

**Pull request merged**